File: scripts/detected_object_3d_labelling.py
#!/usr/bin/env python3
"""
Labelling 3d tracks with 2d sensors
"""
import numpy as np
from collections import Counter, defaultdict, deque
import rclpy
from rclpy.node import Node
from rclpy.time import Time
from rclpy.duration import Duration
from sensor_msgs.msg import CameraInfo
from bb_perception_msgs.msg import (
    DetectedObject2DArray,
    DetectedObject3DArray,
    DetectedObject3D,
    DetectedObject2D,
    ObjectClassification,
)
from tf2_ros.buffer import Buffer
from tf2_ros.transform_listener import TransformListener
from tf2_geometry_msgs import do_transform_pose
from geometry_msgs.msg import PoseStamped
from geometry_msgs.msg import Pose
from std_msgs.msg import Header
from typing import List, Tuple
from shapely.geometry import box
from scipy.optimize import linear_sum_assignment
from copy import deepcopy

np.set_printoptions(formatter={"float": lambda x: "{0:0.3f}".format(x)})


class DetectedObject3DLabelingNode(Node):
    def __init__(self):
        super().__init__("detected_object_3d_labelling_node")

        # Declare parameters
        self.declare_parameter("detection_2d_topic", "/asv4/vision/detections_2d")
        self.declare_parameter(
            "detection_3d_topic", "/asv4/vision/lidar_small_objects/dets_3d/filtered"
        )
        self.declare_parameter(
            "camera_info_topics",
            [
                "/asv4/left_cam/camera_info",
                "/asv4/right_cam/camera_info",
                "/asv4/front_cam/camera_info",
            ],
        )
        self.declare_parameter(
            "output_labeled_topic", "/asv4/vision/lidar_small_objects/dets_3d/labelled"
        )

        # Get parameters
        self.detection_2d_topic = (
            self.get_parameter("detection_2d_topic").get_parameter_value().string_value
        )
        self.detection_3d_topic = (
            self.get_parameter("detection_3d_topic").get_parameter_value().string_value
        )
        self.camera_info_topics = (
            self.get_parameter("camera_info_topics")
            .get_parameter_value()
            .string_array_value
        )
        self.output_labeled_topic = (
            self.get_parameter("output_labeled_topic")
            .get_parameter_value()
            .string_value
        )

        # Set up subscribers
        self.camera_info_dict = {}
        for topic in self.camera_info_topics:
            self.create_subscription(CameraInfo, topic, self.camera_info_callback, 10)

        self.create_subscription(
            DetectedObject2DArray,
            self.detection_2d_topic,
            self.detection_2d_callback,
            10,
        )
        self.create_subscription(
            DetectedObject3DArray,
            self.detection_3d_topic,
            self.detection_3d_callback,
            10,
        )

        self.publisher = self.create_publisher(
            DetectedObject3DArray, self.output_labeled_topic, 10
        )

        # self.inflate_width = 1.5
        self.inflate_width = 0.3
        self.inflate_height = 0.6
        self.bbox_inflate_factor = 1.5
        self.latest_3d_detections = (None, None)
        self.timeout_duration = Duration(seconds=5.0)
        self.track_identities = defaultdict(Counter)

        self.tf_buffer = Buffer(Duration(seconds=15), self)
        self.tf_listener = TransformListener(self.tf_buffer, self, spin_thread=True)

    def camera_info_callback(self, camera_info: CameraInfo):
        self.camera_info_dict[camera_info.header.frame_id] = camera_info

    def detection_3d_callback(self, detection_3d_msg: DetectedObject3DArray):
        self.latest_3d_detections = (
            Time.from_msg(detection_3d_msg.header.stamp),
            detection_3d_msg,
        )

    def detection_2d_callback(self, detection_2d_msg: DetectedObject2DArray):
        # expire 3d detections
        if self.latest_3d_detections[0] is not None and (
            Time.from_msg(detection_2d_msg.header.stamp) - self.latest_3d_detections[0]
            > self.timeout_duration
        ):
            self.latest_3d_detections = (None, None)
        if self.latest_3d_detections[0] is None:
            return
        next_dets = self.latest_3d_detections[1]
        camera_info = self.camera_info_dict.get(detection_2d_msg.sensor.frame_id)
        if not camera_info:
            self.get_logger().warn(
                f"No CameraInfo for frame {detection_2d_msg.sensor.frame_id}"
            )
            return

        labeled_3d_objects = DetectedObject3DArray()
        labeled_3d_objects.header = next_dets.header
        labeled_3d_objects.name = next_dets.name
        labeled_3d_objects.source = next_dets.source
        labeled_3d_objects.sensor_pose = next_dets.sensor_pose

        cost_matrix = (
            np.ones((len(next_dets.objects), len(detection_2d_msg.objects))) * 1e9
        )
        # self.get_logger().info(f"Num objects: {len(next_dets.objects)} {len(detection_2d_msg.objects)}")
        for i, obj_3d in enumerate(next_dets.objects):
            projected_2d_points, dist = self.project_3d_to_2d(
                camera_info,
                detection_2d_msg.header,
                detection_2d_msg.sensor_pose,
                obj_3d,
            )

            if not projected_2d_points or dist < 0:
                continue

            # best_overlap = 0
            # best_class_id = -1

            for j, det_2d in enumerate(detection_2d_msg.objects):
                det_bbox = self.get_bbox_from_2d_detection(det_2d)
                proj_bbox = self.get_bbox_from_2d_points(projected_2d_points)
                overlap = self.compute_overlap(det_bbox, proj_bbox)
                cost_matrix[i][j] = (
                    1 / (overlap + 1e-9) * dist
                )  # prioritize nearer objects

                # if overlap > best_overlap:
                #     best_overlap = overlap
                #     best_class_id = det_2d.hypothesis.class_id

            # if best_overlap > 0.02:
            #     self.track_identities[obj_3d.hypothesis.track_id][best_class_id] += 1
        if min(cost_matrix.shape) == 0:
            return
        if cost_matrix.min() > 1 / (0.01 + 1e-9):
            return
        assignments = linear_sum_assignment(cost_matrix)

        for row, col in zip(assignments[0], assignments[1]):
            track_id = next_dets.objects[row].hypothesis.track_id
            class_id = detection_2d_msg.objects[col].hypothesis.class_id
            self.track_identities[track_id][class_id] += 1

        # Label 3D objects based on the highest count
        for obj_3d in next_dets.objects:
            obj = deepcopy(obj_3d)
            track_counts = self.track_identities[obj.hypothesis.track_id]
            if track_counts:
                most_common_class, count = track_counts.most_common(1)[0]
                second_most_common = (
                    sorted(track_counts.values(), reverse=True)[1]
                    if len(track_counts) > 1
                    else 0
                )
                if count > 3 and count > 1.2 * second_most_common:
                    obj.hypothesis.class_id = most_common_class
                else:
                    self.get_logger().info(
                        f"top label not good enough {count > 3} {count} {second_most_common}",
                        throttle_duration_sec=2.0,
                    )
                    obj.hypothesis.class_id = 0

                total_counts = sum(track_counts.values())
                class_map = defaultdict(float)
                for k, v in track_counts.items():
                    class_map[k] += v / total_counts
                total_prob_sum = sum(class_map.values())
                for k, v in class_map.items():
                    obj.hypothesis.classes.append(
                        ObjectClassification(class_id=k, score=v / total_prob_sum)
                    )
            labeled_3d_objects.objects.append(obj)
        labeled_3d_objects.header.stamp = detection_2d_msg.header.stamp
        self.publisher.publish(labeled_3d_objects)

    def get_bbox_from_2d_detection(self, detection: DetectedObject2D) -> box:
        x = detection.centre_x
        y = detection.centre_y
        width = detection.bbox_width * self.bbox_inflate_factor
        height = detection.bbox_height * self.bbox_inflate_factor
        return box(x, y, x + width, y + height)

    def get_bbox_from_2d_points(self, points: List[Tuple[float, float]]) -> box:
        xs, ys = zip(*points)
        return box(min(xs), min(ys), max(xs), max(ys))

    def project_3d_to_2d(
        self,
        camera_info: CameraInfo,
        header: Header,
        sensor_pose: Pose,  # sensor pose is the camera's position and orientation in the world frame
        obj_3d: DetectedObject3D,
    ):
        # Extract the camera intrinsics
        fx = camera_info.p[0]
        fy = camera_info.p[5]
        cx = camera_info.p[2]
        cy = camera_info.p[6]

        # Extract the 3D object's centroid position
        # x_3d = obj_3d.hypothesis.kinematics.pose_with_covariance.pose.position.x
        # y_3d = obj_3d.hypothesis.kinematics.pose_with_covariance.pose.position.y
        # z_3d = obj_3d.hypothesis.kinematics.pose_with_covariance.pose.position.z
        transformed_pose = None
        try:
            # Lookup the transform from the 3D object's frame to the sensor (camera) frame
            transform = self.tf_buffer.lookup_transform(
                camera_info.header.frame_id,
                obj_3d.hypothesis.kinematics.header.frame_id,
                rclpy.time.Time(),
                Duration(seconds=0.1),
            )
            # self.get_logger().info(f"Transform lookup success: {transform} {camera_info.header.frame_id} {obj_3d.hypothesis.kinematics.header.frame_id}")

            # Create a PoseStamped object for transformation
            pose_stamped = PoseStamped()
            pose_stamped.header = obj_3d.hypothesis.kinematics.header
            pose_stamped.pose = obj_3d.hypothesis.kinematics.pose_with_covariance.pose

            transformed_pose = do_transform_pose(pose_stamped.pose, transform)

        except Exception as e:
            self.get_logger().warn(f"Transform lookup failed: {e}")
            return [], -1

        # Check if the object is in front of the camera
        if transformed_pose.position.z <= 0:
            return [], -1

        # self.get_logger().info(f"pt: {point_in_sensor_frame} {header.frame_id}")

        # Project the 3D point to 2D
        u = (transformed_pose.position.x * fx / transformed_pose.position.z) + cx
        v = (transformed_pose.position.y * fy / transformed_pose.position.z) + cy

        # Calculate the 2D bounding box size based on the distance from the camera
        bbox_width = (
            (
                max(
                    obj_3d.hypothesis.shape.dimensions.x,
                    obj_3d.hypothesis.shape.dimensions.y,
                )
                + self.inflate_width
            )
            / transformed_pose.position.z
            * fx
        )
        bbox_height = (
            (obj_3d.hypothesis.shape.dimensions.z + self.inflate_height * 2)
            / transformed_pose.position.z
            * fy
        )
        bbox_width = max(bbox_width, 5)
        bbox_height = max(bbox_height, 5)

        # Define the corners of the 2D bounding box
        bbox_corners = [
            (u - bbox_width / 2, v - bbox_height / 2),  # Bottom-left corner
            (u + bbox_width / 2, v - bbox_height / 2),  # Bottom-right corner
            (u + bbox_width / 2, v + bbox_height / 2),  # Top-right corner
            (u - bbox_width / 2, v + bbox_height / 2),  # Top-left corner
        ]

        return bbox_corners, np.linalg.norm(
            [transformed_pose.position.x, transformed_pose.position.y]
        )

    def compute_overlap(self, bbox1: box, bbox2: box) -> float:
        intersection = bbox1.intersection(bbox2).area
        union = bbox1.union(bbox2).area
        return intersection / union if union > 0 else 0


def main(args=None):
    rclpy.init(args=args)
    node = DetectedObject3DLabelingNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/detected_object_3d_labelling.cpp
#include <memory>
#include <vector>
#include <unordered_map>
#include <string>
#include <utility>

#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/camera_info.hpp"
#include "bb_perception_msgs/msg/detected_object2_d_array.hpp"
#include "bb_perception_msgs/msg/detected_object3_d_array.hpp"
#include "bb_perception_msgs/msg/detected_object2_d.hpp"
#include "bb_perception_msgs/msg/detected_object3_d.hpp"
#include "bb_perception_msgs/msg/object_classification.hpp"
#include "geometry_msgs/msg/pose_stamped.hpp"
#include "geometry_msgs/msg/pose.hpp"
#include "tf2_ros/buffer.h"
#include "tf2_ros/transform_listener.h"
#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"

using std::placeholders::_1;
using bb_perception_msgs::msg::DetectedObject2DArray;
using bb_perception_msgs::msg::DetectedObject3DArray;
using bb_perception_msgs::msg::DetectedObject2D;
using bb_perception_msgs::msg::DetectedObject3D;
using bb_perception_msgs::msg::ObjectClassification;
using sensor_msgs::msg::CameraInfo;
using geometry_msgs::msg::PoseStamped;
using geometry_msgs::msg::Pose;

class DetectedObject3DLabelingNode : public rclcpp::Node {
public:
    DetectedObject3DLabelingNode()
        : Node("detected_object_3d_labelling_node"), tf_buffer_(this->get_clock()), tf_listener_(tf_buffer_) {

        declare_parameter<std::string>("detection_2d_topic", "/asv4/vision/detections_2d");
        declare_parameter<std::string>("detection_3d_topic", "/asv4/vision/lidar_small_objects/dets_3d/filtered");
        declare_parameter<std::vector<std::string>>("camera_info_topics", {"/asv4/left_cam/camera_info", "/asv4/right_cam/camera_info", "/asv4/front_cam/camera_info"});
        declare_parameter<std::string>("output_labeled_topic", "/asv4/vision/lidar_small_objects/dets_3d/labelled");

        get_parameter("detection_2d_topic", detection_2d_topic_);
        get_parameter("detection_3d_topic", detection_3d_topic_);
        get_parameter("camera_info_topics", camera_info_topics_);
        get_parameter("output_labeled_topic", output_labeled_topic_);

        for (const auto & topic : camera_info_topics_) {
            camera_info_subscribers_.emplace_back(create_subscription<CameraInfo>(topic, 10, std::bind(&DetectedObject3DLabelingNode::camera_info_callback, this, _1)));
        }

        detection_2d_subscriber_ = create_subscription<DetectedObject2DArray>(
            detection_2d_topic_, 10, std::bind(&DetectedObject3DLabelingNode::detection_2d_callback, this, _1));

        detection_3d_subscriber_ = create_subscription<DetectedObject3DArray>(
            detection_3d_topic_, 10, std::bind(&DetectedObject3DLabelingNode::detection_3d_callback, this, _1));

        labeled_3d_publisher_ = create_publisher<DetectedObject3DArray>(output_labeled_topic_, 10);

        inflate_width_ = 1.5;
    }

private:
    void camera_info_callback(const CameraInfo::SharedPtr camera_info) {
        camera_info_dict_[camera_info->header.frame_id] = *camera_info;
    }

    void detection_3d_callback(const DetectedObject3DArray::SharedPtr detection_3d_msg) {
        latest_3d_detections_ = detection_3d_msg;
    }

    void detection_2d_callback(const DetectedObject2DArray::SharedPtr detection_2d_msg) {
        if (!latest_3d_detections_) {
            return;
        }

        const auto & camera_info = camera_info_dict_.find(detection_2d_msg->sensor.frame_id);
        if (camera_info == camera_info_dict_.end()) {
            RCLCPP_WARN(this->get_logger(), "No CameraInfo for frame %s", detection_2d_msg->sensor.frame_id.c_str());
            return;
        }

        DetectedObject3DArray labeled_3d_objects;
        labeled_3d_objects.header = latest_3d_detections_->header;
        labeled_3d_objects.name = latest_3d_detections_->name;
        labeled_3d_objects.source = latest_3d_detections_->source;
        labeled_3d_objects.sensor_pose = latest_3d_detections_->sensor_pose;

        std::vector<std::vector<double>> cost_matrix(latest_3d_detections_->objects.size(), std::vector<double>(detection_2d_msg->objects.size(), 1e9));

        for (size_t i = 0; i < latest_3d_detections_->objects.size(); ++i) {
            const auto projected_2d_points = project_3d_to_2d(camera_info->second, detection_2d_msg->header, detection_2d_msg->sensor_pose, latest_3d_detections_->objects[i]);

            if (projected_2d_points.empty()) {
                continue;
            }

            for (size_t j = 0; j < detection_2d_msg->objects.size(); ++j) {
                const auto det_bbox = get_bbox_from_2d_detection(detection_2d_msg->objects[j]);
                const auto proj_bbox = get_bbox_from_2d_points(projected_2d_points);
                double overlap = compute_overlap(det_bbox, proj_bbox);
                cost_matrix[i][j] = 1 / (overlap + 1e-9);
            }
        }

        if (cost_matrix.empty()) {
            return;
        }

        auto [assignments_row, assignments_col] = linear_sum_assignment(cost_matrix);

        for (size_t row = 0; row < assignments_row.size(); ++row) {
            size_t i = assignments_row[row];
            size_t j = assignments_col[row];

            auto & track_id = latest_3d_detections_->objects[i].hypothesis.track_id;
            auto class_id = detection_2d_msg->objects[j].hypothesis.class_id;
            track_identities_[track_id][class_id] += 1;
        }

        label_3d_objects(labeled_3d_objects);
        labeled_3d_publisher_->publish(labeled_3d_objects);
    }

    void label_3d_objects(DetectedObject3DArray & labeled_3d_objects) {
        for (auto & obj_3d : latest_3d_detections_->objects) {
            auto & track_counts = track_identities_[obj_3d.hypothesis.track_id];
            if (!track_counts.empty()) {
                auto most_common_class = std::max_element(track_counts.begin(), track_counts.end(), [](const auto & a, const auto & b) {
                    return a.second < b.second;
                });
                obj_3d.hypothesis.class_id = most_common_class->first;

                labeled_3d_objects.objects.push_back(obj_3d);
            }
        }
    }

    std::vector<std::pair<double, double>> project_3d_to_2d(const CameraInfo & camera_info, const std_msgs::msg::Header & header, const Pose & sensor_pose, const DetectedObject3D & obj_3d) {
        // Perform 3D to 2D projection
        try {
            auto transform = tf_buffer_.lookupTransform(camera_info.header.frame_id, obj_3d.hypothesis.kinematics.header.frame_id, rclcpp::Time(0), rclcpp::Duration(0.1));
            PoseStamped pose_stamped;
            pose_stamped.header = obj_3d.hypothesis.kinematics.header;
            pose_stamped.pose = obj_3d.hypothesis.kinematics.pose_with_covariance.pose;

            Pose transformed_pose = tf2::doTransform(pose_stamped.pose, transform);

            // Only project objects in front of the camera
            if (transformed_pose.position.z <= 0) {
                return {};
            }

            // Compute 2D projection
            double fx = camera_info.p[0];
            double fy = camera_info.p[5];
            double cx = camera_info.p[2];
            double cy = camera_info.p[6];

            double u = (transformed_pose.position.x * fx / transformed_pose.position.z) + cx;
            double v = (transformed_pose.position.y * fy / transformed_pose.position.z) + cy;

            double bbox_width = (std::max(obj_3d.hypothesis.shape.dimensions.x, obj_3d.hypothesis.shape.dimensions.y) + inflate_width_) / transformed_pose.position.z * fx;
            double bbox_height = (obj_3d.hypothesis.shape.dimensions.z + inflate_width_) / transformed_pose.position.z * fy;

            return {
                {u - bbox_width / 2, v - bbox_height / 2},
                {u + bbox_width / 2, v - bbox_height / 2},
                {u + bbox_width / 2, v + bbox_height / 2},
                {u - bbox_width / 2, v + bbox_height / 2}
            };
        } catch (const std::exception & e) {
            RCLCPP_WARN(this->get_logger(), "Failed to lookup transform: %s", e.what());
            return {};
        }
    }

    std::pair<double, double> get_bbox_from_2d_detection(const DetectedObject2D & obj_2d) {
        double xmin = obj_2d.bbox.center.position.x - obj_2d.bbox.size_x / 2;
        double xmax = obj_2d.bbox.center.position.x + obj_2d.bbox.size_x / 2;
        double ymin = obj_2d.bbox.center.position.y - obj_2d.bbox.size_y / 2;
        double ymax = obj_2d.bbox.center.position.y + obj_2d.bbox.size_y / 2;
        return {xmin, xmax, ymin, ymax};
    }

    std::pair<double, double> get_bbox_from_2d_points(const std::vector<std::pair<double, double>> & points) {
        double xmin = std::min_element(points.begin(), points.end(), [](const auto & a, const auto & b) { return a.first < b.first; })->first;
        double xmax = std::max_element(points.begin(), points.end(), [](const auto & a, const auto & b) { return a.first > b.first; })->first;
        double ymin = std::min_element(points.begin(), points.end(), [](const auto & a, const auto & b) { return a.second < b.second; })->second;
        double ymax = std::max_element(points.begin(), points.end(), [](const auto & a, const auto & b) { return a.second > b.second; })->second;
        return {xmin, xmax, ymin, ymax};
    }

    double compute_overlap(const std::pair<double, double> & bbox1, const std::pair<double, double> & bbox2) {
        double intersect_width = std::max(0.0, std::min(bbox1.second, bbox2.second) - std::max(bbox1.first, bbox2.first));
        double intersect_height = std::max(0.0, std::min(bbox1.fourth, bbox2.fourth) - std::max(bbox1.third, bbox2.third));
        double intersection_area = intersect_width * intersect_height;
        double bbox1_area = (bbox1.second - bbox1.first) * (bbox1.fourth - bbox1.third);
        double bbox2_area = (bbox2.second - bbox2.first) * (bbox2.fourth - bbox2.third);
        double union_area = bbox1_area + bbox2_area - intersection_area;
        return intersection_area / union_area;
    }

    // Fields
    std::string detection_2d_topic_;
    std::string detection_3d_topic_;
    std::vector<std::string> camera_info_topics_;
    std::string output_labeled_topic_;
    double inflate_width_;
    DetectedObject3DArray::SharedPtr latest_3d_detections_;

    rclcpp::Subscription<DetectedObject2DArray>::SharedPtr detection_2d_subscriber_;
    rclcpp::Subscription<DetectedObject3DArray>::SharedPtr detection_3d_subscriber_;
    rclcpp::Publisher<DetectedObject3DArray>::SharedPtr labeled_3d_publisher_;

    std::vector<rclcpp::Subscription<CameraInfo>::SharedPtr> camera_info_subscribers_;
    std::unordered_map<std::string, CameraInfo> camera_info_dict_;

    std::unordered_map<int32_t, std::unordered_map<int32_t, int>> track_identities_;
    tf2_ros::Buffer tf_buffer_;
    tf2_ros::TransformListener tf_listener_;
};

int main(int argc, char ** argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<DetectedObject3DLabelingNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
